git fake data functions git_add, git_commit

waylonho opened the pull request:
parameterized git_add and git_commit helper functions

added author and email parameters for git_commit

carlosparadis commented on DESCRIPTION:
```diff
@@ -43,4 +43,4 @@ Imports:
 URL: https://github.com/sailuh/kaiaulu
 BugReports: https://github.com/sailuh/kaiaulu/issues
-RoxygenNote: 7.0.2
+RoxygenNote: 7.2.3
```
You should not "git add" files which are not needed for the change. This is side effect behavior! This commit should be reverted to not modify this file.

carlosparadis commented on R/git.R:
```diff
@@ -154,31 +154,51 @@ git_create_sample_log <- function(folder_path="/tmp"){
+  #' @param folder_path The worktree path
+  #' @param file_to_add The file we want to add
+  #' @export
+  git_add <- function(git_repo, folder_path, file_to_add) {
```
the function name is already git add, so lets just call the parameter `file`. Can the `filepath` be provided rather than just the file name? if so, then we should call it `filepath` instead!

carlosparadis commented on R/git.R:
```diff
@@ -154,31 +154,51 @@ git_create_sample_log <- function(folder_path="/tmp"){
   # git --git-dir sample/.git --work-tree sample add hello.R
+  #' @param git_repo The git repo path
+  #' @param folder_path The worktree path
+  #' @param file_to_add The file we want to add
```
Good job on including the docs! 👍

carlosparadis commented on conf/geronimo.yml:
```diff
@@ -58,8 +58,8 @@ issue_tracker:
     # Download using `download_jira_data.Rmd`
-    issues: ../../rawdata/issue_tracker/geronimo_issues.json
-    issue_comments: ../../rawdata/issue_tracker/geronimo_issue_comments.json
+    issues: ../rawdata/issue_tracker/geronimo_issues.json
+    issue_comments: ../rawdata/issue_tracker/geronimo_issue_comments.json
```
This is another file that should not be modified. I also recommend you organize the folders as kaiaulu/kaiaulu so you do not need to keep switching the `../` back and forth every time.

carlosparadis commented on man/git_create_sample_log.Rd:
Thanks for including the .Rd files!

carlosparadis reviewed (changes requested):
There are some files being modified on accident here for this PR. Please see my inline comments!

carlosparadis commented:
Note your commit did not pass the test coverage checks, although @Rubegen did! You should compare notes on why yours are not passing, and check the log generated above. 

I suspect it may be some of the files you may have change on accident. 

Please do not delete the fork as you try to change the code! This will kill the Pull Request, and we don't want to create one PR to every attempt :)

carlosparadis commented on vignettes/download_jira_data.Rmd:
````diff
@@ -49,7 +48,7 @@ save_path_issue_tracker_issue_comments <- conf[["issue_tracker"]][["jira"]][["is
 In the project configuration file, we define the domain which JIRA is hosted. The example configuration file uses the domain for Apache Software Foundation. We can use this information to queue the domain to identify all project keys available in its JIRA:
 
-```{r eval = FALSE}
+```{r}
````
You should use git diff before sending the commit so you can spot these. Note you are modifying the original Notebook on parts that should not be modified. This is likely from when I showed how the Notebook works.  Regardless, there should be no reason for the Notebook to be added to your commit, since you are not changing that, right?

carlosparadis commented on R/git.R:
```diff
@@ -154,31 +154,51 @@ git_create_sample_log <- function(folder_path="/tmp"){
+
+
+  # calling functions
+  git_add(git_repo, folder_path, "hello.R")
+  git_commit(git_repo, folder_path, "This is a commit msg.", "John Smith", "test@email.com")
```
This is not a script file. It should not execute code outside function definitions.

carlosparadis reviewed (changes requested):
Still need a few more changes. Consider using git diff before sending a commit for me to review.

carlosparadis commented:
You should also do a git squash at this point to see if the tests goes through.

waylonho commented:
Not sure why this is failing. I've edited the commits so it only changes the git_add and git_commit, and it returns git_repo.

carlosparadis commented:
If you inspect the log it will say. It is failing a unit test:

<img width="949" alt="Screen Shot 2023-10-13 at 6 01 00 AM" src="https://github.com/sailuh/kaiaulu/assets/17270563/725511a4-ff73-4505-8a1a-22db7970a29e">

Which is:

https://github.com/sailuh/kaiaulu/blob/cb2f4098d6aeedb1cf3721a4de75803fe97145f1/tests/testthat/test-parser.R#L52-L53

Did you run the unit tests and check command on your local machine?

waylonho commented:
Been trying to trouble shoot this for a while. Even when I run the check on the master branch on my local machine, I get the following error: 

![91e290cbb503879bc7ff9d0098b8740a](https://github.com/sailuh/kaiaulu/assets/84744515/c852c166-8e8a-47e8-8e28-e42cbccb53a8)

Looking at the R-CMD-check, this line always brings up the error:

Error ('test-parser.R:52:3'): Calling parse_gitlog with correct perceval and correct git log path returns a data table ──
Error in `eval(bysub, x, parent.frame())`: object 'data.Author' not found

It is the same error as mentioned on #108. Not sure if this is a file path issue or something else. Any thoughts?

carlosparadis commented:
Hi, 

As we talked about it today, the error you are experiencing on the pull request is likely due to modifications you performed on git_sample_gitlog. Since one of the functions were incorrectly implemented, it is likely git_sample_log is also broken. If the data can't be generated, then parse_gitlog has no data to test, leading to failure.

Executing the unit tests on your local machine will not work, because you have not installed Perceval.

https://github.com/sailuh/kaiaulu/blob/cb2f4098d6aeedb1cf3721a4de75803fe97145f1/.github/workflows/R-CMD-check.yml#L61C11-L61C157

We did the pip on call. What was missing is knowing where the perceval binary was installed in your machine. It should be in the same place as GitHub Actions install it. I.e. it should be on  /Library/Frameworks/Python.framework/Versions/Current/bin/perceval

If not, just search for the word "perceval" on your machine.

Please let me know the soonest if you can find this file or not. Thanks!

waylonho commented:
Hi. I see a perceval file in usr/local/bin and also one in my python3.10/dist-packages folder. Not sure if this is where it's supposed to go. I've attached an image:

![758dad413297a69813ab568518eda8ac](https://github.com/sailuh/kaiaulu/assets/84744515/f921b8ae-259e-4eed-9ffb-532b5672286e)

Thanks,

carlosparadis commented:
@waylonho 

The thing to keep in mind is that it doesn't matter where the file should be, but rather that you specify its location on `tools.yml` :) Try to first open the terminal and run ./perceval. See if they work. Also try `./perseval -v` and see it is version 0.12.4. 

Once you confirm that, get the full path and replace on my `tools.yml` on Kaiaulu repo with your path. Then try to run parse gitlog function. That should do the trick. Let me know how it goes!

waylonho commented:
Slightly losing my mind but I will figure this out. I have a way too new version installed. When trying to install 0.12.4, for some reason it's not showing the version. Thought it was peculiar, will look into it again.

![d6e4e5a81cd73148d44652fc1f1935db](https://github.com/sailuh/kaiaulu/assets/84744515/a33765f0-8a70-4c61-8fe3-bce1edc0d0e4)

carlosparadis commented:
@waylonho Did we not install 0.12.4 on call? In any case: Try to run Perceval on a `.git` folder. It does not need to be 0.12.4 to work, but prior group had some issues with more recent versions. If it can parse and generate a table, then you should be fine.

carlosparadis commented on R/git.R:
```diff
@@ -125,60 +125,121 @@ git_blame <- function(git_repo_path,flags,commit_hash,file_path){
+#' @param filepath The filepath we want to add
+#' @return The path to the sample .git file.
+#' @export
+git_commit <- function(git_repo, folder_path, commit_msg, author, email) {
```
Seems the unit tests are now passing, but Check is failing. This seems to be the cause per the log:

<img width="565" alt="Screen Shot 2023-10-25 at 10 10 53 PM" src="https://github.com/sailuh/kaiaulu/assets/17270563/5e2df77a-afe6-48d3-9b64-b2e2678be5c9">

carlosparadis commented on R/git.R:
```diff
@@ -125,60 +125,121 @@ git_blame <- function(git_repo_path,flags,commit_hash,file_path){
-  # git --git-dir sample/.git --work-tree sample add hello.R
+  git_add(git_repo, folder_path, file_path)
+
+  # WILL UNCOMMENT THIS BELOW LINE AFTER TEST
```
Should remove these before sending the Pull Request, or they may be accidentally accepted into the main code.

carlosparadis reviewed (changes requested):
I didn't exhaustive look at the code on this review, but wanted to point the more immediate PR failed check.

Also, please take a look on CONTRIBUTING.md to go through the checklist, these changes should be added to NEWS.md under documentation sub-section. 

Thanks!

carlosparadis commented:
@waylonho remember to add yourself to the DESCRIPTION file and add a note on NEWS.md too on this PR. See the repo CONTRIBUTING.md for what to edit there. Let me know when this is ready for review by hitting the request for review button on GitHub.

waylonho commented:
@carlosparadis I will update the DESCRIPTION and NEWS.md shortly. Everything else works as intended, EXCEPT for this one test case:

https://github.com/sailuh/kaiaulu/blob/90be27ee0a3917d330656203b89e31bf891f4057/tests/testthat/test-parser.R#L73-L82

Specifically, line 78 still brings up the "object 'data.Author' not found" error. For the sake of passing the tests, I commented out the lines. I am unsure of how to fix it and believe it's some way the parse_gitlog is trying to call a table from an empty_repo. I will update the files when I can get the test to pass. Thanks.

carlosparadis commented:
@waylonho Your code is a few commits behind from main branch since I had to update Kaiaulu in between, which is leading to the DESCRIPTION and NEWS.md conflict. To address this, you have to git pull master, then git merge to your changes, and finally commit again to your fork branch associated to this PR. 

However, this does not block me from code reviewing. Are the checks passing locally on your machine? Just wanted a final OK before I download your code to review.

carlosparadis commented:
Actually let me try and resolve them via GitHub directly. Standby.

carlosparadis commented:
@waylonho I commit to your fork to resolve the conflicts. The checks are now able to run on the PR. Please make sure to git pull from **your fork** to add my resolving conflict commit before you make more code changes, or you will be unable to make further commits to the PR unless you git clone your fork again.

Let's see what the GitHub checks comes up with!

carlosparadis commented:
@waylonho Unfortunately my edit was ill formatted on the DESCRIPTION file:

```
* checking DESCRIPTION meta-information ... ERROR
Error: Malformed Authors@R field:
  <text>:14:13: unexpected symbol
13:     person('Waylon', 'Ho', role = c('ctb)),
Error: Error in proc$get_built_file() : Build process failed
14:     person('Nicole
Calls: <Anonymous> ... build_package -> with_envvar -> force -> <Anonymous>
                ^
```

It is probably missing a comma or a parenthesis somewhere (your commit was missing Nicole's). Since I can't commit directly to your fork, you need to patch this DESCRIPTION file and send another commit so the checks can go through. Let me know if this doesn't make sense.

carlosparadis commented:
Ok it seems I can commit directly to this branch on your fork. The unit test failing is because of a new test I added tracking another problem. Will commit here in a bit after I fix it on the master branch to see what we get. The description file is now fixed.

carlosparadis commented:
There we go. The checks now pass. I will proceed to code review later today. Make sure you run `git pull origin 227-setup-gitadd-gitcommit-helper-functions` before you make any changes local, or you wind up unable to edit anything.  

Until code review for this is ready, I suggest you, after doing the git pull to this one so you don't forget, focus on the milestone 2. 

On your computer, this means you have to 1) git checkout to your master branch, 2) update your master branch with kaiaulu more recent changes:

* git remote add upstream git@github.com:sailuh/kaiaulu.git
* git pull upstream master
* git merge upstream/master master 

Resolve any conflicts and commit if needed (it shouldn't, unless you accidentally changed code here). 

Then it is business as usual as you do CONTRIBUTING.md: git checkout -b new_branch etc.

carlosparadis reviewed (changes requested):
As asked on @Rubegen PR, could you ensure that

1) the git.R of both your PRs are exact the same
2) the examples.R only contain your examples
3) the unit tests only contain your unit tests

I want to avoid reviewing your incomplete git.R functions or the same code twice in case it ended up duplicated. This also should facilitate showcasing your contributions when the PR is accepted. 

Thanks!

waylonho commented:
Hi, sorry I'm writing this so late. The most recent commit has everything updated:

git.R has all the functions, git_checkout, and git_create_sample_log

io.R has all the functions and is the same as Rubens commit.

I sent Ruben a copy and will update him in the morning.

carlosparadis commented:
@waylonho 

Ruben commit already got merged, so you need to incorporate his code into yours! The simplest way to do this if you don't want to deal with git branches:

Clone the more recent copy of Kaiaulu (not Ruben's) master branch. Add your needed modifications to git.R, io.R, example.R, test-parser.R. Then copy all these files and paste to your branch of this PR. Then send as commit. (Remember to also keep your NEWS.md and DESCRIPTION files.

waylonho commented:
Hi, I currently have all the files done. Just want to make sure: do you want the PR to include Ruben's recent changes that were merged to master, or just my changes? For example, example.R has both our tests?

carlosparadis commented:
@waylonho I see the point of conflict! Since I already merged @waylonho, please go ahead and merge everything from `master` branch to yours, so the conflicts are resolved. The original request was indeed to have your two codes separate, since I was not expecting to have the chance to review either of yours and merge to master. There were a number of other small changes to `git.R` and `io.R` so it is easier at this point you just bring your PR up to date so I can merge!

carlosparadis commented:
@waylonho 

```
  Functions or methods with usage in documentation object 'example_different_branches' but not in code:
    ‘example_different_branches’
  
  Functions or methods with usage in documentation object 'example_different_files_commits' but not in code:
    ‘example_different_files_commits’
  
  Functions or methods with usage in documentation object 'example_empty_repo' but not in code:
    ‘example_empty_repo’
  
  Functions or methods with usage in documentation object 'git_rename' but not in code:
    ‘git_rename’
  
  Functions or methods with usage in documentation object 'io_create_folder' but not in code:
    ‘io_create_folder’
  
  Functions or methods with usage in documentation object 'io_make_sample_file' but not in code:
    ‘io_make_sample_file’
```

I suspect the issue are your .Rd files. Try to generate the docs and version them. I am seeing above the old name of functions that I fixed on @Rubegen code. You may want to wait for the checks to finish before sending the next review request just in case too, unless you are stuck!

carlosparadis commented:
the error generated is because you are missing one function I added to @Rubegen code (Error in ``io_create_folder("kaiaulu_sample"): could not find function "io_create_folder"` I am also a bit worried the NAMESPACE is deleting `git mv` and re-adding `git rename`. This is going in the wrong direction :( 

As I said before, it may be easier to just to just backup your code, copy all of kaiaulu/master from sailuh git.R, io.R, example.R, and test.R, and then re-add your changes on top of it.

carlosparadis commented on R/git.R:
```diff
@@ -201,12 +179,9 @@ git_delete_sample_log <- function(folder_path="/tmp"){
-
-
-#' Git Init
+#' Initialize a git repo in a folder.
```
This is reverting the changes on master. Can you leave the code on master as is?

carlosparadis commented on R/git.R:
```diff
@@ -201,12 +179,9 @@ git_delete_sample_log <- function(folder_path="/tmp"){
 #'
-#' Initializes a new Git repository in the specified folder.
-#' The Git Init command creates a hidden `.git` folder.
+#' This function initializes a new Git repository in the specified folder.
```
Same here.

carlosparadis commented on R/git.R:
```diff
@@ -216,16 +191,13 @@ git_init <- function(folder_path) {
 }
 
-#' Git Commit
+#' Git commit
```
Also reverting the function from master branch.

carlosparadis commented on R/git.R:
```diff
@@ -216,16 +191,13 @@ git_init <- function(folder_path) {
 #'
-#' Executes the `git commit` command on the target
-#' git repository.
+#' Git commits a file
```
Same here.

carlosparadis commented on R/git.R:
```diff
@@ -250,10 +222,10 @@ git_commit <- function(git_repo, folder_path, commit_msg, author, email) {
 
-#' Git Add
+
+#' Git adds a file to the gitlog sample
```
Again reverting, etc.

carlosparadis commented on R/git.R:
```diff
@@ -8,22 +8,38 @@
 #' @return Any error message generated by git
 #' @export
-git_checkout <- function(commit_hash,git_repo_path){
+git_checkout <- function(commit_hash,git_repo_path,new_branch = FALSE){
```
the changes to git_checkout look needed. I would keep this on your commit.

carlosparadis reviewed (changes requested):
Very minor changes needed I believe.

Your changes to git.R are reverting the corrections I already did to them on @waylonho code, with I believe the exception of the `git_checkout` function. So basically, unless you are adding a new function, or improving a new function, can you keep the one I already reviewed on master? Pasting the master functions to your code should fix this, but you need to make sure the example functions are properly using the corrected functions.

waylonho commented:
Ok, after a few technical difficulties the checks pass now. Everything should be updated and contain Rubens code as well as added changes:

- git.R has an updated git_create_sample_log & git_checkout
- example.R has all the test cases
- test-parser.R has all the unit tests
- updated all calls to io functions to proper names (io_make_folder and io_make_file) and added extra folder_path parameter to io_make_folder calls
- updated DESCRIPTION, NEWS.md should be the same (Ruben's new feature addition looked sufficient)

And yes, I was attempting to do the clone master and update that branch, but I was running into trouble because github was changing some other files in the 227 branch, other errors, etc. Long story short I updated 227 branch from master, then added to that branch.

carlosparadis commented:
@waylonho see my review of your passed checks. You're almost there!

waylonho commented:
@carlosparadis Understood, I'll fix these very shortly.

waylonho commented:
Hi, I've commited the original function headers from master, the changes should be good to merge. Let me know if there's any other changes.

carlosparadis commented:
@waylonho please have a look on [a37ccd](https://github.com/sailuh/kaiaulu/pull/232/commits/a374ccdf33a5d21df93c2088bb4d02e34275880a) to see a few things you missed, so you don't repeat on the following PR. 

That being said, good work! The merges were properly addressed. There were a few docs inconsistencies and the suppressWarnings that were left in the code, so remember to not use them on the next run! :^) 

With this, milestone 1 comes to an end! Thank you for your hard work 👍
